File: returned/today1/valid_sudoku.py
# https://leetcode.com/problems/valid-sudoku/
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for i in range(9):
            check_row = set()
            check_col = set()
            for j in range(9):
                if board[i][j] in check_row:
                    return False
                if board[i][j] != '.':
                    check_row.add(board[i][j])
                
                if board[j][i] in check_col:
                    return False
                if board[j][i] != '.':
                    check_col.add(board[j][i])
        
        for gr in range(3):
            for gc in range(3):
                check_grid = set()
                for r in range(3):
                    for c in range(3):
                        cell = board[gr*3+r][gc*3+c]
                        if cell in check_grid:
                            return False
                        if cell != '.':
                            check_grid.add(cell)
        return True

File: typescript/solved/valid-sudoku.ts
/**
 * https://leetcode.com/problems/valid-sudoku/description/
 */
export function isValidSudoku(board: string[][]): boolean {
  return rowsValid(board) && columnsValid(board) && threeByThreeValid(board);
}

export function rowsValid(board: string[][]): boolean {
  for (let row = 0; row < 9; row++) {
    const set = new Set<string>();
    for (let col = 0; col < 9; col++) {
      const val = board[row][col];
      if (val !== '.') {
        if (set.has(val)) return false;
        else set.add(val);
      }
    }
  }

  return true;
}

export function columnsValid(board: string[][]): boolean {
  for (let col = 0; col < 9; col++) {
    const set = new Set<string>();
    for (let row = 0; row < 9; row++) {
      const val = board[row][col];
      if (val !== '.') {
        if (set.has(val)) return false;
        else set.add(val);
      }
    }
  }

  return true;
}

export function threeByThreeValid(board: string[][]): boolean {
  for (let sectionRow = 0; sectionRow <= 2; sectionRow++) {
    for (let sectionCol = 0; sectionCol <= 2; sectionCol++) {
      const rowOffset = sectionRow * 3;
      const set = new Set<string>();
      for (let row = 0 + rowOffset; row < 3 + rowOffset; row++) {
        const colOffset = sectionCol * 3;
        for (let col = 0 + colOffset; col < 3 + colOffset; col++) {
          const val = board[row][col];
          if (val !== '.') {
            if (set.has(val)) return false;
            else set.add(val);
          }
        }
      }
    }
  }

  return true;
}
